plugins/blueprint.cpp

```diff
@@ -592,179 +612,200 @@ command_result do_transform(DFCoord start, DFCoord end, string name, uint32_t ph
+            if (options.auto_phase || options.query)
                 query << "#" << endl;
-            if (phases & PLACE)
+            if (options.auto_phase || options.place)
                 place << "#" << endl;
-            if (phases & BUILD)
+            if (options.auto_phase || options.build)
                 build << "#" << endl;
-            if (phases & DIG)
+            if (options.auto_phase || options.dig)
                 dig << "#" << endl;
         }
-        if (z < end.z - 1)
+        if (z != end.z - z_inc)
         {
-            if (phases & QUERY)
-                query << "#<" << endl;
-            if (phases & PLACE)
-                place << "#<" << endl;
-            if (phases & BUILD)
-                build << "#<" << endl;
-            if (phases & DIG)
-                dig << "#<" << endl;
+            if (options.auto_phase || options.query)
+                query << z_key << endl;
+            if (options.auto_phase || options.place)
+                place << z_key << endl;
+            if (options.auto_phase || options.build)
+                build << z_key << endl;
+            if (options.auto_phase || options.dig)
+                dig << z_key << endl;
         }
     }
-    if (phases & QUERY)
+    if (options.auto_phase || options.query)
         query.close();
-    if (phases & PLACE)
+    if (options.auto_phase || options.place)
         place.close();
-    if (phases & BUILD)
+    if (options.auto_phase || options.build)
         build.close();
-    if (phases & DIG)
+    if (options.auto_phase || options.dig)
         dig.close();
+
     return CR_OK;
 }
 
-bool cmd_option_exists(vector<string>& parameters, const string& option)
+static bool get_options(blueprint_options &opts,
+                        const vector<string> &parameters)
+{
+    auto L = Lua::Core::State;
+    color_ostream_proxy out(Core::getInstance().getConsole());
+    Lua::StackUnwinder top(L);
+
+    if (!lua_checkstack(L, parameters.size() + 2) ||
+        !Lua::PushModulePublic(
+            out, L, "plugins.blueprint", "parse_commandline"))
+    {
+        out.printerr("Failed to load blueprint Lua code");
+        return false;
+    }
+
+    Lua::Push(L, &opts);
+
+    for (const string &param : parameters)
+        Lua::Push(L, param);
+
+    if (!Lua::SafeCall(out, L, parameters.size() + 1, 0))
+        return false;
+
+    return true;
+}
+
+static bool do_gui(const vector<string> &parameters)
+{
+    auto L = Lua::Core::State;
+    color_ostream_proxy out(Core::getInstance().getConsole());
+    Lua::StackUnwinder top(L);
+
+    if (!lua_checkstack(L, parameters.size() + 1) ||
+        !Lua::PushModulePublic(out, L, "plugins.blueprint", "do_gui"))
+    {
+        out.printerr("Failed to load blueprint Lua code");
+        return false;
+    }
+
+    for (const string &param : parameters)
+        Lua::Push(L, param);
+
+    if (!Lua::SafeCall(out, L, parameters.size(), 0))
+        return false;
+
+    return true;
+}
```

Comment:
Is your goal with this function to launch `gui/blueprint` safely? If so, the poorly-named `Core::setHotkeyCmd()` should do the trick too. `autogems.cpp` and `hotkeys.cpp` use it, for instance. I do see that `do_gui()` introduces a 1-frame delay, though - I'm not sure what purpose that serves.

Reply:
Thank you for the pointer to `setHotkeyCmd()` I.. never would have thought to look there ; P

the idea behind the 1-frame delay (and it might not even achieve this goal) was to let the plugin call exit before I launched another related command. I haven't given any thought to re-entrancy, and I didn't want to create a code path that could conceivably call the blueprint plugin while the blueprint plugin was still processing an existing call. For example, if someone ran `blueprint gui --cursor=10,10,10` and immediately sent another `Enter` key to the df ui, it would immediately call blueprint again on the same frame. again, I haven't really analyzed the code behavior in this situation, but I figured I could head off any potential issues by just not making it possible.

Reply:
I made the change, but I'm curious about what you think about the possibility of having two blueprint calls on the stack.

Reply:
to clarify this a bit, I don't think there's a specific path in `blueprint` or `gui/blueprint` that would cause a problem (though I can't guarantee that there will *never* be a problematic path). I'm just not sure about the plugin infrastructure and I'm wary of reentrancy in general in code that was not designed to be multithreaded.

Reply:
> the idea behind the 1-frame delay (and it might not even achieve this goal) was to let the plugin call exit before I launched another related command

I _thought_ `setHotkeyCmd()` handled this by queueing up the command to be executed on the next tick. Looking over it again, it definitely executes commands on a different thread, at least - it's intended to handle hotkeys bound with the `keybinding` command, which are detected in `SDL_PollEvent()`, where it isn't safe to invoke commands directly.

Reply:
I tried making `gui/blueprint` call a `blueprint` command immediately when run, and that seems to run consistently after the first call to `blueprint()` has returned. Seems that there's some delay, at least, even if not a full tick.

Reply:
if it's executed in a different thread, reentrancy wouldn't be an issue.. but then we have the spectre of concurrency. Are there any safeguards against a hotkey command executing while an interactive command is running? Although C++-11 has thread-safe initialization of local statics, I've seen global static variables sprinkled around the code with no mutex guards.

I think this PR is probably good with its usage of `setHotkeyCmd()`, but I'll file a bug for an audit/migration of global statics.

Reply:
> I tried making `gui/blueprint` call a `blueprint` command immediately when run, and that seems to run consistently after the first call to `blueprint()` has returned. Seems that there's some delay, at least, even if not a full tick.

that's reassuring. I'll still file the bug about global statics, though, since it still seems like a good idea to clean those up

Reply:
I _thought_ that hotkey commands ran with a CoreSuspender, but it looks like I might have been wrong about that. Maybe it's the responsibility of the command itself to do that. Maybe just moving the CoreSuspender creation earlier in blueprint() would give more complete thread-safety.

Reply:
good idea. done.